Run history

csordasmarton opened the pull request:
Closes #783

whisperity commented on api/report_server.thrift:
```diff
@@ -32,6 +32,13 @@ struct RunData {
 typedef list<RunData> RunDataList
 
+struct RunHistoryData {
+  1: string tag,      // tag of the report
```
Yet again, format the comments here as a full string, and below, etc.

whisperity commented on api/report_server.thrift:
```diff
@@ -188,6 +197,9 @@ service codeCheckerDBAccess {
+  RunHistoryDataList getRunHistory(1: i64 runId)
+                                   throws (1: shared.RequestFailed requestError),
+
   ReportData getReport(
```
The comment line for what permission is needed to `getReport()` went missing as the new function was introduced above.

whisperity commented on libcodechecker/libhandlers/store.py:
```diff
@@ -107,6 +107,13 @@ def add_arguments_to_parser(parser):
+                        dest="tag",
+                        required=False,
+                        default=argparse.SUPPRESS,
+                        help="Tag which identifies a certain run update.")
```
Okay... there are a lot of concepts not really explained here. "Update"? "Tag"? I think this help message should be made better...

I believe `An identifier for this individual store of results in the run's history.` (Run history is used on the GUI.)

whisperity commented on libcodechecker/libhandlers/store.py:
```diff
@@ -107,6 +107,13 @@ def add_arguments_to_parser(parser):
                              "the '--name' parameter given to 'codechecker-"
                              "analyze' will be used, if exists.")
 
+    parser.add_argument('--tag',
```
A new argument is added, but the documentation for the `store` command was not updated.

whisperity commented on libcodechecker/server/run_db_model.py:
```diff
@@ -70,6 +70,25 @@ def mark_finished(self):
+                    ForeignKey('runs.id', deferrable=True,
+                               initially="DEFERRED", ondelete='CASCADE'),
+                    index=True)
+    tag = Column(String)
```
Tags need not be unique then?

csordasmarton replied:
If it would be unique, how would I add the same tag for the same or for a different run?
E.g:

| run_id | tag | user | date |
| --------- | ---- | ------ | ------ |
1	| v1.0 | Anonymous | 2017-09-06 16:56:33.681858
1	| v1.0 | Anonymous | 2017-09-06 17:56:33.681858

The question is, do we want to allow the user to do this?

whisperity replied:
Let's ask the others. @dkrupp @gyorb Depends on whether we want to consider this annotation as a _tag_ or as a _comment_. Tags, such as version tags, are usually unique.

dkrupp replied:
the tag should be unique at least within a run.

gyorb replied:
Maybe the name using the name `version tag` would be more obvious.

whisperity commented on libcodechecker/server/run_db_model.py:
```diff
@@ -198,6 +217,9 @@ class Report(Base):
                                    'reopened',
                                    name='detection_status'))
 
+    detected_at = Column(DateTime)
```
I believe a `NOT NULL` constraint should exist on the `detected_at` column? Every report is expected to have at least a detection date.

whisperity commented on libcodechecker/server/client_db_access_handler.py:
```diff
@@ -11,6 +11,7 @@
 import codecs
 from collections import defaultdict
 import datetime
+from datetime import timedelta
```
We only use `timedelta` at one place, I'd rather not have a separate import just for this.

whisperity commented on libcodechecker/server/client_db_access_handler.py:
```diff
@@ -116,6 +117,30 @@ def process_report_filter_v2(report_filter, count_filter=None):
+
+    if report_filter.fixedDate:
+        date = datetime.strptime(report_filter.fixedDate, '%Y-%m-%d')
+        tomorrow = date + timedelta(days=1)
```
Please use `datetime.timedelta()` if it is possible. At other places, `datetime.strfptime()` is used in a similar manner.

csordasmarton replied:
If I'm using this way I'm getting the following error:
```bash
type object 'datetime.datetime' has no attribute 'timedelta'
```
I used the following import to resolve this problem:
```python
from datetime import datetime, timedelta
```

whisperity commented on libcodechecker/server/client_db_access_handler.py:
```diff
@@ -527,6 +553,36 @@ def getRunData(self, run_name_filter):
 
     @timeit
+    def getRunHistory(self, run_id):
+        try:
```
Only users who have access to the product should be able to query run history.

```py
self.__require_access()
```
, before the `try` statement.

whisperity commented on libcodechecker/server/client_db_access_handler.py:
```diff
@@ -2300,10 +2357,15 @@ def massStoreRun(self, name, version, b64zip, force):
                                                      None)
 
+        user = self.__auth_session.user \
+            if self.__auth_session else "Anonymous"
```
`'Anonymous'`

Xazax-hun commented on api/report_server.thrift:
```diff
@@ -94,7 +101,9 @@ struct ReportFilter_v2 {
   6: list<shared.ReviewStatus>    reviewStatus,
-  7: list<shared.DetectionStatus> detectionStatus
+  7: list<shared.DetectionStatus> detectionStatus,
+  8: optional string              detectedDate,
```
Isn't it better to store something like an ID to the run in the history where it was first detected/fixed?

whisperity replied:
Since CodeChecker 6, a run ID is a very, very long running stable variable. A run is only introduced once and then for the entire lifespan of the project, it is not expected to be deleted, only updated (stored into) later on. Storing does not change the run ID.

whisperity replied:
Though I would like it better if dates would be sent as at least integers (in epoch time), not strings, and let the client do the formatting of these values (perhaps different clients with region-specific formats?)

Xazax-hun replied:
I did not mean the run ID, but an ID to an entry in the run history table.

csordasmarton commented:
I also did some minor changes: in case of the enum-based filters (severity, review status), I encoded the string value (instead of the enum value) into the URL.
```
http://localhost:8001/Default/#run=2&detection-status=resolved&severity=high
```

csordasmarton commented:
All travis test cases are successfully passed.

gyorb commented on api/v6/report_server.thrift:
```diff
@@ -25,6 +25,22 @@ struct RunData {
 typedef list<RunData> RunDataList
 
+struct RunHistoryData {
+  1: i64       runId,    // Unique id of the run.
```
Do we need to return the `runId` the client calls the api with a runId `getRunHistory(1: i64 runId)`?

csordasmarton replied:
Yes, we need it, because we can call the `getRunHistory` function without run id. In this case we get back history data for all runs. (See Run history tab under All reports).

gyorb commented on api/v6/report_server.thrift:
```diff
@@ -25,6 +25,22 @@ struct RunData {
+struct RunHistoryData {
+  1: i64       runId,    // Unique id of the run.
+  2: string    runName,  // Name of the run.
+  3: string    tag,      // Tag of the report.
```
Shouldn't this be the tag of the run?
Will we support only one tag for each history data entry?

csordasmarton replied:
- A version tag is uniquely identifying a store. Actually, the last tag at the run history data is the tag of the run.
- We support only one tag for each history data as I know.

gyorb commented on api/v6/report_server.thrift:
```diff
@@ -94,7 +110,10 @@ struct ReportFilter_v2 {
-  7: list<shared.DetectionStatus> detectionStatus
+  7: list<shared.DetectionStatus> detectionStatus,
+  8: list<string>                 runHistoryTag,
+  9: optional string              detectedDate,
```
I think `detectionDate` or `firstDetectionDate` would be more expressive. A comment would be nice that this value is set only once.

gyorb commented on api/v6/report_server.thrift:
```diff
@@ -94,7 +110,10 @@ struct ReportFilter_v2 {
+  7: list<shared.DetectionStatus> detectionStatus,
+  8: list<string>                 runHistoryTag,
+  9: optional string              detectedDate,
+  10: optional string             fixedDate
```
How about `fixDate`? I think we should comment here that this date can change.

gyorb commented on docs/user_guide.md:
```diff
@@ -668,6 +668,8 @@ optional arguments:
                         reports to the database. If not specified, the '--
                         name' parameter given to 'codechecker-analyze' will be
                         used, if exists.
+  --tag TAG             An identifier for this individual store of results in
```
A more descriptive help message would be better to note to the user that an already used tag can not be used again.

gyorb commented on libcodechecker/server/client_db_access_handler.py:
```diff
@@ -49,6 +50,7 @@ class CountFilter:
     SEVERITY = 3
     REVIEW_STATUS = 4
     DETECTION_STATUS = 5
+    RUN_HISTORY_TAG = 5
```
This should be `6`.

gyorb commented:
If I check the `all reports` tab and I select one tag in the filters I get the number of reports but the list of reports is empty.

Going from the all reports tab and run history if I select one tagged version the run id will be shown as the run name and the run counts will show wrong numbers.

gyorb commented:
Products page is not shown correctly, I'm not sure because of the changes here.

gyorb commented:
I can reproduce the product listing problem with chrome, firefox works fine for me.

csordasmarton commented:
- I fixed the error message of an already existing tag for the run to this:
```bash
[14:50] - Storing analysis results for run 'tinyxml-resolved'
API call error: massStoreRun
RequestFailed(errorCode=2, _message='Tag v1.0 is already used in this run')
[14:50] - Storage failed: RequestFailed(errorCode=2, _message='Tag v1.0 is already used in this run')
```
- I fixed the empty report error.
- I created an issue for product listing error (#912).

whisperity commented on www/style/codecheckerviewer.css:
```diff
@@ -938,3 +963,95 @@ span[class*="severity-"] {
+  font-size: 1.5em;
+  font-weight: bold;
+  color: #000;
+  font-family: Times, serif;
```
Is this style here intentional? We don't really use _serif_ fonts in the GUI. Looks awkward.

gyorb commented:
After cleaning all the filter values setting the `Fix date` (only the day) in the `Detection date` filter the request to the server crashes.
